### ISSUE TITLE
"Shipping address is not set" error for virtual product added to cart before login

### ISSUE THREAD

**Issue by pguedesbr** [magento/magento2#8039]
### Preconditions

1. Magento 2.1.3
2. PHP 5.6.25
3. MySQL 5.6.23

### Steps to reproduce

1. Have a virtual product (may be configurable, as long as the child product is virtual)
2. Logout of the customer account
3. Add one virtual product to the cart
4. Go to checkout
5. Fill in the customer email (of an existing customer account)
6. Wait for it to recognize as an existing customer, fill in the password and login

![image](https://cloud.githubusercontent.com/assets/9370105/21658074/3209b0a2-d2ac-11e6-99e6-35489d46d3a3.png)

7. Select Check/money order as payment method
8. Select one of the existing billing address (there's no shipping address since the product is virtual) anc click Update button

![image](https://cloud.githubusercontent.com/assets/9370105/21658140/7c38aef8-d2ac-11e6-9132-a4c1990f4485.png)

### Expected result

1. Billing address should save (address select field turns into address text block)
2. User would click on Place Order button and be redirected to the success page

### Actual result

1. The payment method section changes the radio button + "check/money order" to "No payment information required"

2. The sidebar with cart items renders empty items and order total changes to $0.00

![image](https://cloud.githubusercontent.com/assets/9370105/21658327/4c162204-d2ad-11e6-8273-c06a07c67c43.png)

2. Upon clicking on Place Order button, theres the error: "Shipping address is not set"

![image](https://cloud.githubusercontent.com/assets/9370105/21658493/0854611a-d2ae-11e6-9743-d6366c62c1ca.png)

### Extra observations

I believe somehow magento looses track of the cart items added before login. When we choose billing address after login, the sidebar reloads with no items and $0.00 (if you reload the page the items are still there. Its not like the items were removed from the cart)

By loosing information about the items, Magento may not detect its a virtual order (since there are no virtual products in the cart) therefore requiring shipping address to place the order (even thought the shipping step was skipped in the checkout due to it actually being a virtual order)

Regarding the "No payment information required", this may also be related to the order totals becoming $0.00 for the time being

### Other use cases

- If you checkout as a guest, all is good
- if you login before adding the products to the cart, all is good
- If you login after adding the products, but then add another [virtual] product after logging in, all is good

Long story short:
- Checkout with all items added before login: ERROR
- Checkout with at least one item added after login: OK

**Comment by abarua17** [comment]
@pguedesbr I have not been able to reproduce the issue. I created a virtual product. I created a customer with two address. I logged out from the customer account. I added a virtual product to the cart. Then on checkout, added the email address and it asked me to enter password. After that I selected Check/Money order. I selected the billing address from dropdown. It turned into an address block with the address that i had selected. Then I clicked on place order. I was successfully able to do so.  Do you have any other additional information regarding your environment setting?

**Comment by pguedesbr** [comment]
Hello, the steps I've posted on the original post are the ones with which i can reproduce the error all the time.

Im running in developer mode, block and full page caches are disabled.

Right now I've triggered the same "sidebar cart item contents changing when i confirm billing address", which seems related as Ive mentioned before.

1. Logged in as a customer, I've added 2 items to the cart
2. Logged out as a customer (cart becomes empty when you logout)
3. Add 1 item to cart (cart now has 1 item)
4. Log back in as a customer (despite the 2 items previously in the cart before logout, it remains with only one item, the one added last)
5. Go to checkout (checkout page sidebar shows the only item in cart)
6. Since you're already logged in, just choose one address for billing (check/money order)
7. Click on "update button" to confirm/submit selected billing address
8. After address turns from <select> into text block, sidebar updates and switches the cart content from the only product to the 2 other products

Looks like somehow it is retrieving old cart items from the database in this last step. If you refresh the page though, checkout restart with the 1 item again.

**Comment by pguedesbr** [comment]
I'm closing this issue since i found an error in my module code which was causing an exception during quote merge (anonymous + customer quotes) after login.
